### PR TITLE
Echo Reduction

### PR DESCRIPTION
**Pull request by wchmzn**
**Issue #:**

**Description of changes:**
Add Amazon Chime Echo Reduction for React SDK
**Testing**

1. How did you test these changes?
Testing steps:
- Enable Web Audio. Enable Echo Reduction. Join the meeting
- Select one input audio device, and then click on "Enable voice focus".
- Expect: ns_es model get download and exactly once.
2. Can these changes be tested using one of the demo application? If yes, which demo application can be used to test it?
apps/meeting
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by jing-chen1 — COMMENTED** [review]
I would suggest to add some screen shots in the description since it involves some UI changes.

File: apps/meeting/serverless/src/package.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "chime-sdk-meeting-serverless-demos",
+  "version": "0.1.0",
+  "description": "Amazon Chime SDK Serverless Demos",
+  "dependencies": {
+    "aws-sdk": "^2.1034.0"
```
[the remark this answers is not in the document]
Comment (wchmzn) [reply apps/meeting/serverless/src/package.json:6]: i wan't aware of adding this in the beginning but looks like without this package.json aws-sdk would fall back to an old default version.

File: apps/meeting/server.js [hunk 2]
```diff
@@ -16,6 +16,15 @@ let options = {};
 
 const chime = new AWS.Chime({ region: 'us-east-1' });
 const alternateEndpoint = process.env.ENDPOINT;
+
+const chimeRegional = new AWS.ChimeSDKMeetings({ region: 'us-east-1' });
```
[the remark this answers is not in the document]
Comment (wchmzn) [reply apps/meeting/server.js:20]: sure

File: apps/meeting/server.js [hunk 3]
```diff
@@ -16,6 +16,15 @@ let options = {};
...
+const chimeRegionalEndpoint = process.env.REGIONAL_ENDPOINT || 'https://meetings-chime.us-east-1.amazonaws.com';
+
+if (chimeRegionalEndpoint) {
```
[the remark this answers is not in the document]
Comment (wchmzn) [reply apps/meeting/server.js:23]: removed checkings

File: apps/meeting/server.js [hunk 4]
```diff
@@ -51,21 +74,31 @@ const server = require(protocol).createServer(
...
-          meetingCache[title] = await chime
-            .createMeeting({
-              ClientRequestToken: uuidv4(),
-              MediaRegion: region
-            })
-            .promise();
+          let request = {
+            ClientRequestToken: uuidv4(),
+            MediaRegion: region,
+            ExternalMeetingId: title.substring(0, 64),
```
[the remark this answers is not in the document]
Comment (wchmzn) [reply apps/meeting/server.js:82]: try to remove but got this error `Server error: Missing required key 'ExternalMeetingId' in params`

File: apps/meeting/server.js [hunk 5]
```diff
@@ -51,21 +74,31 @@ const server = require(protocol).createServer(
...
+          };
+          if (query.ns_es === 'true') {
```
[the remark this answers is not in the document]
Comment (wchmzn) [reply apps/meeting/server.js:84]: it is from String(echoReductionCapability). default is string false

File: apps/meeting/server.js [hunk 6]
```diff
@@ -51,21 +74,31 @@ const server = require(protocol).createServer(
         const name = query.name;
         const region = query.region || 'us-east-1';
 
+        let client = getClientForMeeting(meetingCache[title]);
```
[the remark this answers is not in the document]
Comment (wchmzn) [reply apps/meeting/server.js:77]: moved query.ns_es to getClientForMeeting

File: apps/meeting/serverless/deploy.js [hunk 7]
```diff
@@ -141,6 +142,8 @@ if (!fs.existsSync('build')) {
 console.log(`Using region ${region}, bucket ${bucket}, stack ${stack}, disable-printing-logs ${disablePrintingLogs}`);
 ensureBucket();
 
+spawnOrFail('npm', ['install'], {cwd: path.join(__dirname, 'src')});
```
[the remark this answers is not in the document]
Comment (wchmzn) [reply apps/meeting/serverless/deploy.js:145]: removed

File: apps/meeting/src/containers/MeetingProviderWrapper/index.tsx [hunk 8]
```diff
@@ -78,9 +79,32 @@ const MeetingProviderWrapper: React.FC = () => {
     );
   };
 
+  function voiceFocusName(name: string | undefined): VoiceFocusModelName | undefined {
```
[the remark this answers is not in the document]
Comment (wchmzn) [reply apps/meeting/src/containers/MeetingProviderWrapper/index.tsx:82]: removed undefined

File: apps/meeting/src/providers/AppStateProvider.tsx [hunk 9]
```diff
@@ -17,16 +17,20 @@ interface AppStateValue {
...
   layout: Layout;
+  joinInfo: any;
   toggleTheme: () => void;
   toggleWebAudio: () => void;
   toggleSimulcast: () => void;
   togglePriorityBasedPolicy: () => void;
   setBlurValue: (blurValue: string) => void;
+  toggleEchoReduction: () => void;
   setMeetingMode: React.Dispatch<React.SetStateAction<MeetingMode>>;
+  setJoinInfo: (joinInfo: any) => void;
```
[the remark this answers is not in the document]
Comment (wchmzn) [reply apps/meeting/src/providers/AppStateProvider.tsx:33]: added

File: apps/meeting/serverless/src/handlers.js [hunk 10]
```diff
@@ -110,14 +130,25 @@ exports.createMeeting = async (event, context, callback) => {
   }
 
   let meetingInfo = await getMeeting(title);
+  let client = getClientForMeeting(meetingInfo);
   if (!meetingInfo) {
-    const request = {
+    let request = {
```
[the remark this answers is not in the document]
Comment (wchmzn) [reply apps/meeting/serverless/src/handlers.js:135]: changed

File: apps/meeting/src/containers/MeetingProviderWrapper/index.tsx [hunk 11]
```diff
@@ -78,9 +79,32 @@ const MeetingProviderWrapper: React.FC = () => {
     );
   };
 
+  function voiceFocusName(name: string | undefined): VoiceFocusModelName {
```
Comment (jing-chen1) [inline apps/meeting/src/containers/MeetingProviderWrapper/index.tsx:82]: Do we need `undefined` here?

File: apps/meeting/src/containers/MeetingProviderWrapper/index.tsx [hunk 12]
```diff
@@ -78,9 +79,32 @@ const MeetingProviderWrapper: React.FC = () => {
...
+    if (name && ['default', 'ns_es'].includes(name)) {
+      return name as VoiceFocusModelName;
+    }
+    return 'default';
+  }
+
+  function getVoiceFocusSpecName(): VoiceFocusModelName | undefined {
```
Comment (jing-chen1) [inline apps/meeting/src/containers/MeetingProviderWrapper/index.tsx:89]: Do we need this `undefined` as return type?

File: apps/meeting/src/providers/AppStateProvider.tsx [hunk 13]
```diff
@@ -17,16 +18,20 @@ interface AppStateValue {
   region: string;
   isWebAudioEnabled: boolean;
   blurOption: string;
+  isEchoReductionEnabled: boolean;
   meetingMode: MeetingMode;
   enableSimulcast: boolean;
   priorityBasedPolicy: VideoPriorityBasedPolicy | undefined;
   layout: Layout;
+  joinInfo: JoinMeetingInfo| undefined;
```
Comment (jing-chen1) [inline apps/meeting/src/providers/AppStateProvider.tsx:26]: [nit]: `JoinMeetingInfo| undefined` -> `JoinMeetingInfo | undefined`

File: apps/meeting/src/providers/AppStateProvider.tsx [hunk 14]
```diff
@@ -17,16 +18,20 @@ interface AppStateValue {
...
   toggleTheme: () => void;
   toggleWebAudio: () => void;
   toggleSimulcast: () => void;
   togglePriorityBasedPolicy: () => void;
   setBlurValue: (blurValue: string) => void;
+  toggleEchoReduction: () => void;
   setMeetingMode: React.Dispatch<React.SetStateAction<MeetingMode>>;
+  setJoinInfo: (joinInfo: JoinMeetingInfo| undefined) => void;
```
Comment (jing-chen1) [inline apps/meeting/src/providers/AppStateProvider.tsx:34]: [nit]: `joinInfo: JoinMeetingInfo| undefined` -> `joinInfo: JoinMeetingInfo | undefined`

File: apps/meeting/src/providers/AppStateProvider.tsx [hunk 15]
```diff
@@ -54,12 +59,14 @@ export function AppStateProvider({ children }: Props) {
   const [meetingId, setMeetingId] = useState(query.get('meetingId') || '');
   const [region, setRegion] = useState(query.get('region') || '');
   const [meetingMode, setMeetingMode] = useState(MeetingMode.Attendee);
+  const [joinInfo, setJoinInfo] = useState<JoinMeetingInfo| undefined>(undefined);
```
Comment (jing-chen1) [inline apps/meeting/src/providers/AppStateProvider.tsx:62]: [nit]: `JoinMeetingInfo| undefined` -> `JoinMeetingInfo | undefined`

File: apps/meeting/src/types/index.ts [hunk 16]
```diff
@@ -50,4 +50,4 @@ export const BlurValues = {
   blur10Percent: '10',
   blur20Percent: '20',
   blur40Percent: '40',
-};
+};
```
Comment (jing-chen1) [inline apps/meeting/src/types/index.ts:53]: [nit]: new line at the end of file

File: apps/meeting/src/utils/api.ts [hunk 17]
```diff
@@ -5,11 +5,21 @@ import routes from '../constants/routes';
...
+  Meeting: CreateMeetingResponse;
+  Attendee: string;
+}
+
 interface MeetingResponse {
-  JoinInfo: {
-    Attendee: string;
-    Meeting: string;
-  };
+  JoinInfo?: JoinMeetingInfo;
```
Comment (jing-chen1) [inline apps/meeting/src/utils/api.ts:22]: Do we want to add `?` here

**Pull request closed without merging**